File: js/signal/src/PeerManager.ts
import type { Peer, SignalPresenceData } from './types';

/**
 * Maps actorTokenId ↔ Peer, filtering self.
 */
export class PeerManager {
  private _peers = new Map<string, Peer>();
  private _actorToPeerId = new Map<string, string>();
  private _localActorId: string;

  constructor(localActorId: string) {
    this._localActorId = localActorId;
  }

  /**
   * Add or update a peer from presence data.
   * Returns the Peer if it's a remote peer, null if it's self.
   */
  addFromPresence(actorTokenId: string, presence: SignalPresenceData, joinedAt?: number): Peer | null {
    const isLocal = actorTokenId === this._localActorId;

    // Skip self
    if (isLocal) return null;

    const existing = this._actorToPeerId.get(actorTokenId);
    const peerId = presence.peerId || existing || actorTokenId;

    const peer: Peer = {
      peerId,
      actorTokenId,
      connectionState: 'new',
      metadata: presence.metadata,
      joinedAt: joinedAt || Date.now(),
      isLocal: false,
    };

    this._peers.set(peerId, peer);
    this._actorToPeerId.set(actorTokenId, peerId);

    return peer;
  }

  /**
   * Remove a peer by actorTokenId.
   * Returns the removed peer, or null if not found / is self.
   */
  removeByActorId(actorTokenId: string): Peer | null {
    if (actorTokenId === this._localActorId) return null;

    const peerId = this._actorToPeerId.get(actorTokenId);
    if (!peerId) return null;

    const peer = this._peers.get(peerId) || null;
    this._peers.delete(peerId);
    this._actorToPeerId.delete(actorTokenId);

    return peer;
  }

  /**
   * Get a peer by peerId.
   */
  getPeer(peerId: string): Peer | undefined {
    return this._peers.get(peerId);
  }

  /**
   * Get a peer by actorTokenId.
   */
  getPeerByActorId(actorTokenId: string): Peer | undefined {
    const peerId = this._actorToPeerId.get(actorTokenId);
    return peerId ? this._peers.get(peerId) : undefined;
  }

  /**
   * Get all remote peers.
   */
  getAll(): Peer[] {
    return Array.from(this._peers.values());
  }

  /**
   * Get the peers Map (readonly view).
   */
  get peers(): Map<string, Peer> {
    return this._peers;
  }

  /**
   * Clear all tracked peers.
   */
  clear(): void {
    this._peers.clear();
    this._actorToPeerId.clear();
  }
}

File: python/signal/nolag_signal/peer_manager.py
from __future__ import annotations

import time
from typing import Any

from .types import Peer, SignalPresenceData


class PeerManager:
    """Bidirectional map of actorTokenId <-> Peer, filtering self."""

    def __init__(self, local_actor_id: str) -> None:
        self._local_actor_id = local_actor_id
        self._peers: dict[str, Peer] = {}  # peerId -> Peer
        self._actor_to_peer_id: dict[str, str] = {}  # actorTokenId -> peerId

    def add_from_presence(
        self,
        actor_token_id: str,
        presence: dict[str, Any],
        joined_at: float | None = None,
    ) -> Peer | None:
        if actor_token_id == self._local_actor_id:
            return None

        peer_id = presence.get("peerId", actor_token_id)
        metadata = presence.get("metadata")

        peer = Peer(
            peer_id=peer_id,
            actor_token_id=actor_token_id,
            connection_state="new",
            metadata=metadata,
            joined_at=joined_at or time.time(),
            is_local=False,
        )

        self._peers[peer_id] = peer
        self._actor_to_peer_id[actor_token_id] = peer_id
        return peer

    def remove_by_actor_id(self, actor_token_id: str) -> Peer | None:
        if actor_token_id == self._local_actor_id:
            return None

        peer_id = self._actor_to_peer_id.pop(actor_token_id, None)
        if peer_id is None:
            return None
        return self._peers.pop(peer_id, None)

    def get_peer(self, peer_id: str) -> Peer | None:
        return self._peers.get(peer_id)

    def get_peer_by_actor_id(self, actor_token_id: str) -> Peer | None:
        peer_id = self._actor_to_peer_id.get(actor_token_id)
        if peer_id is None:
            return None
        return self._peers.get(peer_id)

    def get_all(self) -> list[Peer]:
        return list(self._peers.values())

    def clear(self) -> None:
        self._peers.clear()
        self._actor_to_peer_id.clear()
